### ISSUE TITLE
Opening multiple files of the same type in the same qube for editing can cause data loss

### ISSUE THREAD

**Issue by UndeadDevel** [QubesOS/qubes-issues#8931]
### Qubes OS release
4.2
4.1 as well IIRC

### Brief summary
Using a named DVM as standard target for `qvm-open-in-dvm` allows opening multiple files in the same DVM (see #814, #5690).
The problem is that if these files are of the same file type such that they will be opened in that DVM by the same program (tested for Geany and LibreOffice Writer), ~~then some files in `/tmp/` will presumably be overwritten,~~ such that in the case where one might want to not just view but also edit multiple files in that "DVM session", the saves will not necessarily go through, which is handled differently by different programs, which typically don't make it clear that saving data will now have no effect, as the saved data won't get pushed back to the origin qube.

This cannot happen with unnamed DVMs due to #814 and #5690, but it does also happen for non-DVM qubes.

So far I could only reproduce when opening files of the same type; editing multiple files of different types, such that they will be opened by different programs in the target DVM doesn't cause this issue so far.

### Steps to reproduce
1. Create a named DVM and use it as target for the following `qvm-open-in-dvm` ops
2. Create multiple files of the same type (e.g. odt or txt) in an AppVM
3. Open multiple of them in the named DVM and edit some of them
4. Save the changes and observe error message from program (if any)
5. Close the program and check if changes were pushed back to origin AppVM

### Expected behavior
All saved changes are pushed back to the origin qube.

### Actual behavior
In above described scenario this often fails and changes are not pushed back to origin qube without obvious error message, which can lead to data loss in such scenarios.

**Comment by UndeadDevel** [comment]
Okay I had a closer look at `/tmp` in the target DVM: 
When the first file of the same type (odt) is opened, a new `qrexec-rpc-stderr` pair is created there as well as a file `OSL_PIPE_1000_SingleOfficeIPC_[hexcode]` and a folder with the actual document.
When the second file of the same type (odt) is opened, only a new folder is created, which contains a lock file, but not the document itself; there are no additional pipes created as far as I can see and in the origin qube I see that the cache file for the second document seems to be briefly created but then deleted again; so it looks like something "aborts" the "session" of the second file and deletes it in the target qube, while also deleting the cache file in the origin qube, but the document is still open in the target qube, it just doesn't have a backing in tmpfs anymore (hence the error when trying to save) and any saved changes will create a new file, which won't be pushed back to the origin qube.

**Comment by marmarek** [comment]
> so it looks like something "aborts" the "session" of the second file and deletes it in the target qube

I think something like this is happening: when you have your application open already, starting it the second time (for the second file) doesn't wait for the editing to finish (it opens the file in already running application), and the process exits immediately. This makes `qubes.OpenInVM` service think you already finished editing so it checks if file was modified (it wasn't yet) to decide whether new version needs sending back, and then it removes the file. It happens quickly, so when you look at it, the file is already gone.

You can see it for yourself: in that running DispVM, in a terminal, start the application with some file - you will see it will return to shell immediately.

I don't have a good solution for this, I'm afraid it means editing multiple files in the same disposable qube (or any qube for that matter) is not yet supported by qvm-open-in-vm and requires a bit more work than just adjusting policy.

**Comment by unman** [comment]
You are right, and it depends on how the file type is handled in
the disposable.
"it opens the file in already running application" - this is the issue,
and it depends on the application. For libreoffice, every document opens
as window in same application by default.
The only way to circumvent this is to start the program with a different
user profile:
e.g. /usr/lib/libreoffice/program/soffice --nologo -env:UserInstallation=file:///tmp/1 

So, in this case you could edit the libreoffice scripts to
generate a new env directory on each call and use that new directory
when opening the file(s).
That should work.

**Comment by UndeadDevel** [comment]
Thank you! You're right, that was it; the problem with @unman's solution is that the new profile won't have the settings of the DVM template, but that can be remedied by copying over the old profile before launching.

For anyone interested, my concrete steps were (all in the DVM template):
1. `sudo cp /usr/share/applications/libreoffice* /usr/local/share/applications/`
2. `sudo rm /usr/local/share/applications/libreoffice-startcenter.desktop`
3. `find /usr/local/share/applications/ -type f -name libreoffice* -exec sudo sed -i '0,/Exec=libreoffice/{s/Exec=libreoffice/Exec=\/home\/user\/.local\/bin\/libreoffice.sh/}' {} \;`
4. create `/home/user/.local/bin/libreoffice.sh` with following content (and make it executable):
```
#!/bin/sh
FOLDER=/tmp/LibreOffice_$(date | tr ' :' '_')
mkdir $FOLDER
cp -r ~/.config/libreoffice/4/user $FOLDER/
libreoffice $1 --nologo -env:UserInstallation=file://$FOLDER $2
```

As to this issue...I'm still thinking it's a bug, even if named DVMs in the mentioned scenario require a custom RPC policy, because this issue also occurs when opening multiple files of the same file type in the same target AppVM with the default policy.

**Comment by unman** [comment]
use `mktemp -d` to generate the unique env directory.

**Comment by UndeadDevel** [comment]
> use `mktemp -d` to generate the unique env directory.

Thanks for the tip, but I like it more with the time component, as I sometimes want to copy the modified & saved doc without closing it in the DVM and so the time part helps me know which folder the correct file is in if multiple files are open in the DVM (I could just check the file properties with your proposed solution, but I find it more convenient my way).

On second thought, however, though your solution is probably safer e.g. if multiple files were to be opened within one second...okay I've now combined your solution with mine using a template for `mktmp` and edited the post.

**Comment by marmarek** [comment]
> Regardless, would it be acceptable to distinguish these two calls via arguments? E.g.:
> 
>     * `qubes.OpenInVM+view`
> 
>     * `qubes.OpenInVM+edit`

Yes, that should be okay. The service currently ignores the argument, so it would be just a hint for the policy.
